napi_memory_management.md

```diff
@@ -0,0 +1,175 @@
+# Object Lifetime Management
+
+As N-API calls are made, handles to objects in the heap for the underlying
+vm may be returned as `napi_value`s.  These handles must hold the
+objects 'live' until they are no longer required by the native code,
+otherwise the objects could be collected before the native code was
+finished using them.
+
+As object handles are returned they are associated with a
+'scope'. The lifespan for the default scope is tied to the lifespan
+of the native method call. The result is that, by default, handles
+remain valid and the objects associated with these handles will be
+held live for the lifespan of the native method call.   
+
+In many cases, however, it is necessary that the handles remain valid for
+either a shorter or longer lifespan than that of the native method.
+The sections which follow describe the N-API functions than can be used
+to change the handle lifespan from the default.
+
+## Making handle lifespan shorter than that of the native method
+
+It is often necessary to make the lifespan for handles shorter than
+the lifespan of a native method. For example, consider a native method
+that has a loop which iterates through the elements in a large array:
+
+```C
+  for (int i = 0; i < 1000000; i++) {
+    napi_value result;
+    napi_status status = napi_get_element(e object, i, &result);
+    if (status != napi_ok) {
+      break;
+    }
+    // do something with element
+  }
+```
+
+This would result in a large number of handles being created, consuming
+substantial resources. In addition, even though the native code could only
+use the most recent handle, all of the associated objects would also be
+kept alive since they all share the same scope.
+
+To handle this case, N-API provides the ability to establish a new 'scope' to
+which newly created handle will be associated.  Once those handles
```

Comment:
"... newly created handle*s* will be ..." or "... *a* newly created handle will be ..."